### ISSUE TITLE
feat(telemetry): cross-agent prefix overlap and token reuse accounting in gateway and cacheobs

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12312]
## Reopened: wire shape exists, production attribution does not

The prior close landed the intended session wire fields but not their producer. `guardvars.SessionVars` declares `PromptTokens`, `SharedTokens`, and `ReuseRate` (`internal/guardvars/guardvars.go:45-66`), while production `debugSessions` populates none of them (`internal/gateway/debug.go:764-795`).

Native reuse is captured process-globally and by model/tenant in `cacheobs.Default`: `internal/agent/inkernel_planner.go:2131-2189` records exact prompt, cacheable, realized reused, eligibility, model, authenticated tenant, and source after each successful native completion. `cacheobs.LabeledSnapshot` is keyed by model, tenant, and phase (`internal/cacheobs/labeled.go:112-151`), not request/session/trace. Tenant is an authorization boundary and must not be repurposed as a session identity.

The exact request-local values also leave the planner in `agent.Completion.Usage` (`internal/agent/inkernel_planner.go:2227-2232`). There is no bounded session accumulator joining those values to a turnkey request identity.

## Working spine

Define an explicit turnkey session identity contract; after a successful completion, add its exact prompt and cached-token counts to a bounded per-session accumulator; project that accumulator into the existing `SessionVars` fields; serve it through turnkey `/debug/vars`.

## Typed relations

- Coordinates with: #12874
- Coordinates with: #12315
- Coordinates with: #12812

No start blockers.

## Deterministic witness

A device-free test records two turns for session A and one for session B with fixed prompt/cached values, then proves exact `PromptTokens`, `SharedTokens`, and `ReuseRate` per row; aggregate session sums reconcile with the process-global snapshot; absent identity remains explicitly unknown/unavailable; bounded eviction removes only the selected stale identity.

## Scope

Likely files:

- `cmd/fak/up.go`
- one dedicated turnkey session-attribution file
- one focused test file

A `guardvars` shape change is unnecessary unless implementation finds the existing fields insufficient.

## Gold-plating boundary

No physical KVPool page/byte sharing (#12812), counterfactual savings or proof schema (#12315), inferred parent-child lineage, cache optimization, hardware performance claim, or full gateway telemetry migration.

## Definition of done

- [ ] Explicit turnkey session identity has documented unknown/unavailable behavior.
- [ ] Successful native completions accumulate exact request-local prompt and cached-token values into bounded session rows.
- [ ] `/debug/vars` reports exact per-session prompt/shared totals and reuse rate.
- [ ] Session-row sums reconcile with the scoped aggregate used by the witness.
- [ ] Deterministic bounded-lifecycle witness passes.

**Comment by anthony-chaudhary** [comment]
### Accepted bounded census subphase

This subphase tightens only the existing private cross-agent ledger primitive in `internal/gateway/crossagent_reuse.go` and `internal/gateway/crossagent_reuse_test.go`; the full served-path producer, explicit session identity, and parent/peer prefix-origin proof remain required before #12312 can close.

The ledger will keep prompt/shared-token accumulation unchanged while treating its child count as a lower bound with explicit coverage. Known nonempty child traces distinct from the parent count once. Repeated turns from the same child do not inflate the count. Missing child identity increments an explicit unknown-identity observation count and marks census coverage incomplete without inventing a child. A hard ledger-wide identity budget bounds memory; overflow preserves existing token totals and the retained lower bound, marks coverage incomplete, and does not self-evict a live parent to manufacture a smaller exact count. Existing parent eviction and retain pruning release that parent's tracked identities.

Deterministic source witness: repeated same child => lower bound 1; second child => 2; missing identity => tokens retained, unknown count increments, complete=false; identity-budget overflow => tokens/count retained and complete=false; retain pruning reclaims identity budget. No served-path wiring, cache-origin inference, optimization, hardware, or performance claim is included.

**Comment by anthony-chaudhary** [comment]
## Native prefix-origin audit

Source inspection confirms the remaining attribution gap:

- `gateway.(*Server).complete` has the request trace but calls `agent.WithPrefixCacheIdentity(ctx, principal, "")`; authenticated tenant/principal survives, while agent/child identity is empty (`internal/gateway/gateway.go`). Tenant is an authorization scope, not evidence of who produced a cached prefix.
- `radixkv.node` owns KV/snapshot/logit payloads but records no admitting agent or token-origin lineage (`internal/radixkv/radixkv.go`). Scoped lookup returns share scope, physical tier, and matched length; it cannot distinguish a peer-produced tenant hit from the same child's own warmed tenant entry (`internal/radixkv/safekv.go`).
- The native planner reduces lookup to `matched` tokens and exports only `Completion.Usage.PromptTokensDetails.CachedTokens`. Gateway therefore receives no prefix-origin receipt (`internal/agent/inkernel_planner.go`).
- `debugSessionVars.ParentTrace` is explicitly continuation/reset lineage, not spawn lineage, so it cannot supply a parent-child attribution join (`internal/gateway/debug.go`).

The existing controlled witness surface is `TestInKernelSameTenantExactDeviceL1CachedLogitsSkipsRedundantSnapshotClone` in `internal/agent/inkernel_redundant_snapshot_test.go`: agent-1 cold-primes an exact prompt and previously unused agent-2 in the same tenant receives an exact tenant-scope Device-L1 hit with identical output. Its `countingBackend` sets `deviceMemory=true`; this is a deterministic synthetic native-path witness, not physical-device or savings evidence.

A radix edge cannot carry immutable origin honestly. `installSnapshot` can close and replace an existing node's snapshot without changing the token path, so the current payload generation may have a different computation lineage than the edge's first insertion. By contrast, the CPU exact-insert path with an empty suffix retains the existing KV and only updates logits/chunk metadata; it is not a general payload replacement. Any future representation therefore needs ordered token-origin spans attached to each reusable payload generation, with snapshot materializer/producer recorded separately: reuse plus append derives prefix spans from the source receipt and a suffix span from the current computation; exact snapshot replacement replaces the receipt; split clips it; eviction discards it; and restore without integrity-bound provenance is explicitly unknown. This remains a scoping requirement, not an authorized field/API or cache policy. In particular, creating forced agent-private copies after shared hits is not accepted because it changes cache ownership and copy cost. Historical origin, recurring reuse, and one-time cross-agent benefit attribution require a separate reviewed counting contract. Tenant scope, last writer, snapshot producer, and cached-token count must never be substituted for token origin.

The v3 distinct-child census helper subsequently passed its focused test (executor receipt SHA-256 `4ad4a1765e80195804e22834393a0d435b64bf8204b6b072de3b05fbeeac9cdd`; raw log SHA-256 `42f03c671c26d12c65ce3370b4c2b3423b116ffda1a97594806e260155f01e43`). The v4 behavioral-proof revision has passed source review, while its parent RED and fixed-tree GREEN executions remain pending. The helper is still not wired into the served production path. General live per-child attribution requires origin-aware served observations. A controlled isolated A-only prime followed by a genuinely fresh B-first native request can establish causal sharing without that production lineage machinery only when child-to-request-to-native receipts, request order, cache lifetime, and absence of other traffic are integrity-bound. That controlled witness has not run, so no native cross-agent savings are claimed.


### OpenCode correlation Leaf A source candidate

A bounded source-only candidate now adds an explicit `FAK_OPENCODE_REQUEST_CORRELATION=1` `chat.headers` hook to the canonical OpenCode project asset. It runs only for the configured `fak` provider and emits nonblank, control-character-free `X-Fak-OpenCode-Session-Id`, `X-Fak-OpenCode-Message-Id`, and `X-Fak-OpenCode-Agent` values while preserving all existing headers, including case-insensitive matches. These values are client-asserted correlation metadata only; they are not authentication, tenant identity, cache origin, or causal proof.

Frozen normal landing packet: base `13025078471a3382b11165d7c7595c12253055ea`, two-path patch SHA-256 `de8a7aec96ca143238dddcba000e6075db51940586be134107faa9d1b845c8d1`, production SHA-256 `19f19f3870e2367e73ab857681da68831ca165198745bb818522c04ca0ecb8ea`, test SHA-256 `64f682a68d441070fbe235b4eeca8cbe7cfefa2ef1459e0808dcb70827d0be92`. The ignored generated disk asset is byte-identical to the embedded canonical plugin (SHA-256 `7010d3df40f1c174ca4f91311962951038acccdda87126d2b5c8ec0b1097fcb7`) and is excluded from normal landing.

`git diff --check` passed. The intended Node-backed Go fixture covers disabled, non-`fak`, full, partial/invalid, absent metadata, existing-header preservation, and old-hook presence, but no Go test, Node execution, build, model call, or landing has run. This leaf emits at OpenCode ingress only: current guard-to-`fak up` forwarding and actual native receipt capture remain a separate Leaf B prerequisite. #12312 stays open, and no cache reuse or savings claim follows from this source candidate.

**Comment by anthony-chaudhary** [comment]
### Current status

Current fixed-tree focused census witness: **PASS**. The regression fixture has now been corrected to compile against the parent implementation and expose its repeated-child inflation as behavioral RED; that parent RED and normal full validation still require a separately granted execution. Landing, served-path wiring, causal origin evidence, and savings proof remain pending. Earlier lease and execution events below are retained as dated history.

### Census execution deferred by lease fence

The frozen census candidate remains source-reviewed but was not executed in the completed CPU grant:

- Candidate patch SHA-256: `736e4e21adb49c53c358b0b723c60e517ea9901c65413fa8a91d4ea28b8a6007`.
- Frozen source SHA-256: production `682363a9a60a5d1b09501dc80e163d76a9dbc06d56af990ff8c77cb0df6b90aa`; test `3ecd2f203dd4256c78e1cd06cc0caf3d5df3287e3c5b4447f6698f528d0ecad2`.
- Prepared command: `GOMAXPROCS=2 go test -p 1 -timeout 60s ./internal/gateway -run '^TestCrossAgentReuse(Telemetry|DistinctChildCensus)' -count=1 -v`.
- Typed outcome: SKIP before launch. Canonical local and remote `refs/fak/locks/gateway-crossagent-census-12312` both resolved to object `16f09b73c431e769c24c88f3422a2fed65eb0a42`, generation 1, acquired Unix `1789231834`, TTL 3600, expiry 2026-09-12 17:50:34 UTC. That did not leave the full 300-second outer execution budget, and the grant prohibited lease repair or renewal.

No test process was launched and no source or lease was changed. The next step is a normal same-holder renewal followed by a separate focused CPU grant. Served-path wiring and cross-agent origin/savings claims remain outside this helper witness and are still pending.


### Generation 2 readiness recovery

The normal recovery path has established canonical lease generation 2 for the same holder and exact two-file scope: acquired Unix `1789236398`, TTL 3600, ref object `b571f0f8884def299d41a82fe7b58cf478ff2830`. Recovery receipt: local `_scratch/goals/recovery-12312-lease-receipt.json`. Fresh managed inventory reports owner LIVE, lease LIVE, lifecycle READY. The frozen patch and source hashes remain exactly those recorded above. The historical prelaunch SKIP remains valid; the focused census test is ready for a separate explicit CPU grant and has not been launched by this update.


### Focused census execution PASS

After the generation 2 recovery, a separate bounded CPU grant executed the exact focused witness:

```text
GOMAXPROCS=2 go test -p 1 -timeout 60s ./internal/gateway -run '^TestCrossAgentReuse(Telemetry|DistinctChildCensus)' -count=1 -v
```

Result: exit 0, no skips. The three top-level tests passed: `TestCrossAgentReuseTelemetry`, `TestCrossAgentReuseDistinctChildCensus`, and `TestCrossAgentReuseTelemetryEmpty`. The census test's three named subcases also passed: known repeated plus missing identities; oversized identity token retention without identity retention; and budget overflow token retention plus identity-budget reclaim on retain pruning.

Durable executor receipt: local `_scratch/goals/native-mac-enduser-01a09663/issue-12312-census-v3/execution/executor-receipt.json`, SHA-256 `4ad4a1765e80195804e22834393a0d435b64bf8204b6b072de3b05fbeeac9cdd`. Raw log SHA-256: `42f03c671c26d12c65ce3370b4c2b3423b116ffda1a97594806e260155f01e43`. Production and test source hashes remained unchanged at `682363a9a60a5d1b09501dc80e163d76a9dbc06d56af990ff8c77cb0df6b90aa` and `3ecd2f203dd4256c78e1cd06cc0caf3d5df3287e3c5b4447f6698f528d0ecad2`.

An unrelated external Go build began after launch, so this receipt supports functional correctness only and carries no timing or performance claim. The primitive remains unwired into the served path; it proves neither causal prefix origin nor cross-agent savings. Landing and full issue closure remain pending.




### Test-proof correction

The prior test directly referenced fields and identity-budget internals introduced by the fix, so it could not compile against parent `8f2e8b56a8d1675c1e963aea0dfc39ee25d44972` and did not yet provide a behavioral parent RED. The corrected fixture uses compatibility accessors: it reads the new lower-bound/coverage fields when present and falls back to the legacy count when absent. Two turns from the same child therefore produce legacy count 2 against required distinct count 1, while the fixed tree reports 1. Missing identity, oversized identity, budget overflow, token retention, and retain-pruning assertions remain.

Frozen v4 patch SHA-256: `a421f640f9206df854135cf98f4b46653543f22ed87c13cc3d43396beefcfac7`; production remains `682363a9a60a5d1b09501dc80e163d76a9dbc06d56af990ff8c77cb0df6b90aa`; corrected test is `992445b51e798210640e7c59d84ecbea543cc8ea5fcaefeaf457d9d4140db572`. No test/build was run for this correction. Parent behavioral RED, fixed-tree GREEN refresh, normal validation, and landing remain pending separate grants; the helper remains unwired and #12312 stays open.
